### PR TITLE
di: add safe wrapper around LLVMGlobalCopyAllMetadata

### PR DESCRIPTION
**Pull request by alessandrod**
Add MetadataEntries which is a safe wrapper around LLVMGlobalCopyAllMetadata. Before this change we used to leak the returned entries.

### CODE REVIEW

File: src/llvm/di.rs [hunk 1]
```diff
@@ -348,15 +348,41 @@ unsafe fn iter_operands(v: LLVMValueRef) -> impl Iterator<Item = LLVMValueRef> {
...
-unsafe fn iter_metadata_copy(v: LLVMValueRef) -> impl Iterator<Item = (u32, LLVMMetadataRef)> {
-    let mut count = 0;
-    let entries = LLVMGlobalCopyAllMetadata(v, &mut count);
-    (0..count).map(move |index| {
-        (
-            LLVMValueMetadataEntriesGetKind(entries, index as u32),
-            LLVMValueMetadataEntriesGetMetadata(entries, index as u32),
-        )
-    })
+struct MetadataEntries {
```
Comment (vadorovsky) [inline src/llvm/di.rs:351]: Can we move it to `src/llvm/types/di.rs`?

File: src/llvm/di.rs [hunk 2]
```diff
@@ -348,15 +348,41 @@ unsafe fn iter_operands(v: LLVMValueRef) -> impl Iterator<Item = LLVMValueRef> {
...
+    entries: *mut LLVMValueMetadataEntry,
+    count: usize,
+}
+
+impl MetadataEntries {
+    fn new(v: LLVMValueRef) -> Option<Self> {
+        if !unsafe { can_get_all_metadata(v) } {
```
Comment (vadorovsky) [inline src/llvm/di.rs:358]: I would just do

```suggestion
        if !unsafe { !LLVMIsAGlobalObject(v).is_null() || !LLVMIsAInstruction(v).is_null()  } {
```

here inline. Or at least move these small `is_*`/`can_*` functions to `src/llvm/types/di.rs` as well.

**Review by vadorovsky — CHANGES_REQUESTED** [review]

File: src/llvm/di.rs (hunk 1 above)
Comment (alessandrod) [reply src/llvm/di.rs:351]: Not all metadata is DI?

File: src/llvm/di.rs (hunk 2 above)
Comment (alessandrod) [reply src/llvm/di.rs:358]: ack, removing can_get_all_metadata and removing more in a followup to this

File: src/llvm/di.rs (hunk 1 above)
Comment (vadorovsky) [reply src/llvm/di.rs:351]: True. I see that the `LLVMGlobalCopyAllMetadata` belongs to `llvm/lib/IR/Core.cpp`. I would then keep it in `src/llvm/types/ir.rs` (or, eventually, but probably too muc effort - split `ir.rs` into a subdir with `mod.rs` and `core.rs`, where eventual wrappers for other `IR/*` files could have their own files).

My main point is that I would prefer to not mix wrappers (which are currently in the `types` subdir) and DI sanitization logic in the same module.

**Comment by alessandrod** [comment]
Closed as part of https://github.com/aya-rs/bpf-linker/pull/173

**Pull request closed without merging**